File: backend/src/server/task-store.ts
/**
 * SQLite 任务持久化存储 — 翻译自 Python task_store.py。
 *
 * 使用 better-sqlite3 的同步 API，在 Node.js 单线程模型下性能最优。
 * 方法签名遵守 ITaskStore 接口（返回 Promise<T>），同步逻辑由 async 函数自动包装。
 */

import Database, { type Database as DatabaseType } from 'better-sqlite3';
import { mkdirSync } from 'node:fs';
import { dirname } from 'node:path';
import { type ITaskStore, type TaskRecord, TaskStatus } from './app';

/**
 * SQLite 持久化实现。
 *
 * 设计要点：
 * - 所有 JSON 字段（payload, act_json, evidence）原样存取字符串，不做解析假设。
 * - updateTask 动态构建 SET clause，自动追加 updated_at。
 * - listTasks 使用 ORDER BY updated_at DESC 并严格 (limit, offset) 参数顺序。
 */
export class TaskStore implements ITaskStore {
  private db: DatabaseType | null = null;
  private readonly dbPath: string;

  constructor(dbPath: string) {
    this.dbPath = dbPath;
  }

  // ─── Lifecycle ─────────────────────────────────────────────────

  async initialize(): Promise<void> {
    // 确保父目录存在
    if (this.dbPath !== ':memory:') {
      mkdirSync(dirname(this.dbPath), { recursive: true });
    }

    this.db = new Database(this.dbPath);
    this.db.pragma('journal_mode = WAL');
    this.db.pragma('synchronous = NORMAL'); // 极大提升 WAL 模式下的并发写入能力（写盘转交 OS 缓冲区）
    this.db.pragma('busy_timeout = 5000');  // 防止高并发时抛出 SQLITE_BUSY 导致 Node.js 进程奔溃
    this.db.pragma('foreign_keys = ON');

    this.db.exec(`
      CREATE TABLE IF NOT EXISTS tasks (
        task_id    TEXT PRIMARY KEY,
        petition   TEXT NOT NULL,
        status     TEXT NOT NULL,
        result     TEXT,
        bill_state TEXT NOT NULL DEFAULT 'petition',
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL
      )
    `);

    this.db.exec(`
      CREATE TABLE IF NOT EXISTS events (
        id           INTEGER PRIMARY KEY AUTOINCREMENT,
        task_id      TEXT NOT NULL,
        timestamp    TEXT NOT NULL,
        source_agent TEXT NOT NULL,
        action       TEXT NOT NULL,
        emotion      TEXT DEFAULT 'neutral',
        intensity    REAL DEFAULT 0.5,
        payload      TEXT DEFAULT '{}',
        FOREIGN KEY (task_id) REFERENCES tasks(task_id)
      )
    `);

    this.db.exec(`
      CREATE TABLE IF NOT EXISTS acts (
        task_id    TEXT PRIMARY KEY,
        act_json   TEXT NOT NULL,
        created_at TEXT NOT NULL,
        FOREIGN KEY (task_id) REFERENCES tasks(task_id)
      )
    `);

    this.db.exec(`
      CREATE TABLE IF NOT EXISTS verdicts (
        task_id        TEXT PRIMARY KEY,
        constitutional INTEGER NOT NULL,
        ruling         TEXT NOT NULL,
        evidence       TEXT DEFAULT '[]',
        created_at     TEXT NOT NULL,
        FOREIGN KEY (task_id) REFERENCES tasks(task_id)
      )
    `);

    // Bug Fix: 清理僵尸任务 (Zombie Tasks)
    // 如果 Node.js 进程在执行期间意外崩溃 (如 OOM, SIGKILL, 宿主机断电)，
    // 内存中的队列会被清空，但数据库中残留的状态会永远卡在 PENDING 或 RUNNING。
    // 这会导致前端永久判定为“执行中”。在每次启动时，必须将这些悬空任务置为 FAILED。
    const now = new Date().toISOString();
    const info = this.db.prepare(`
      UPDATE tasks 
      SET status = ?, result = ?, updated_at = ?
      WHERE status IN (?, ?)
    `).run(
      TaskStatus.FAILED, 
      'Server unexpectedly terminated during execution (Zombie Task cleaned on startup)', 
      now, 
      TaskStatus.PENDING, 
      TaskStatus.RUNNING
    );

    if (info.changes > 0) {
      console.warn(`[TaskStore] Cleaned up ${info.changes} zombie task(s) on startup.`);
    }
  }

  async close(): Promise<void> {
    if (this.db) {
      this.db.close();
      this.db = null;
    }
  }

  /**
   * 获取已初始化的 DB 实例，未初始化时立即抛出错误。
   */
  private ensureDb(): DatabaseType {
    if (!this.db) {
      throw new Error('TaskStore not initialized. Call initialize() first.');
    }
    return this.db;
  }

  // ─── Tasks CRUD ────────────────────────────────────────────────

  async createTask(taskId: string, petition: string): Promise<void> {
    const db = this.ensureDb();
    const now = new Date().toISOString();

    const stmt = db.prepare(`
      INSERT INTO tasks (task_id, petition, status, result, bill_state, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `);
    stmt.run(taskId, petition, TaskStatus.PENDING, null, 'petition', now, now);
  }

  async getTask(taskId: string): Promise<TaskRecord | null> {
    const db = this.ensureDb();
    const row = db.prepare('SELECT * FROM tasks WHERE task_id = ?').get(taskId) as Record<string, unknown> | undefined;

    if (!row) return null;

    return {
      task_id: row.task_id as string,
      petition: row.petition as string,
      status: row.status as TaskStatus,
      result: (row.result as string) ?? null,
      bill_state: row.bill_state as string,
      created_at: row.created_at as string,
      updated_at: row.updated_at as string,
    };
  }

  /** updateTask 允许更新的列白名单 — 防止 TS 类型擦除后的 SQL 注入 */
  private static readonly UPDATABLE_COLUMNS = new Set(['status', 'result', 'bill_state']);

  async updateTask(
    taskId: string,
    update: Partial<Pick<TaskRecord, 'status' | 'result' | 'bill_state'>>,
  ): Promise<void> {
    const db = this.ensureDb();
    const entries = Object.entries(update).filter(([, v]) => v !== undefined);
    if (entries.length === 0) return;

    const now = new Date().toISOString();
    // 动态构建 SET clause — 只更新明确传入的字段
    const setClauses: string[] = [];
    const values: unknown[] = [];

    for (const [key, val] of entries) {
      // 防御性白名单：阻止 TS 类型擦除后的非法列名注入
      if (!TaskStore.UPDATABLE_COLUMNS.has(key)) {
        throw new Error(`updateTask: illegal column "${key}"`);
      }
      setClauses.push(`${key} = ?`);
      values.push(val);
    }
    setClauses.push('updated_at = ?');
    values.push(now);
    values.push(taskId); // WHERE 子句参数

    const sql = `UPDATE tasks SET ${setClauses.join(', ')} WHERE task_id = ?`;
    db.prepare(sql).run(...values);
  }

  async countTasks(): Promise<number> {
    const db = this.ensureDb();
    const row = db.prepare('SELECT COUNT(*) AS cnt FROM tasks').get() as { cnt: number };
    return row.cnt;
  }

  async listTasks(offset: number, limit: number): Promise<TaskRecord[]> {
    const db = this.ensureDb();
    // 关键：ORDER BY updated_at DESC 保证最新任务在前
    // 参数绑定 (limit, offset) 对齐 SQL 的 LIMIT ? OFFSET ? 占位符
    const rows = db.prepare(
      'SELECT * FROM tasks ORDER BY updated_at DESC LIMIT ? OFFSET ?',
    ).all(limit, offset) as Array<Record<string, unknown>>;

    return rows.map((row) => ({
      task_id: row.task_id as string,
      petition: row.petition as string,
      status: row.status as TaskStatus,
      result: (row.result as string) ?? null,
      bill_state: row.bill_state as string,
      created_at: row.created_at as string,
      updated_at: row.updated_at as string,
    }));
  }

  // ─── Events ────────────────────────────────────────────────────

  async storeEvent(
    taskId: string,
    sourceAgent: string,
    action: string,
    emotion: string,
    intensity: number,
    payload: string,
  ): Promise<void> {
    const db = this.ensureDb();
    const now = new Date().toISOString();

    db.prepare(`
      INSERT INTO events (task_id, timestamp, source_agent, action, emotion, intensity, payload)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `).run(taskId, now, sourceAgent, action, emotion, intensity, payload);
  }

  async storeEventBatch(
    taskId: string,
    eventData: { sourceAgent: string, action: string, emotion: string, intensity: number, payloadStr: string },
    stateChange?: string,
    actJson?: string,
    verdict?: { constitutional: boolean, ruling: string, evidence: string }
  ): Promise<void> {
    const db = this.ensureDb();
    const now = new Date().toISOString();

    // Bug 52 fix: Use SQLite transactions to prevent partial writes
    const tx = db.transaction(() => {
      // 1. insert event
      db.prepare(`
        INSERT INTO events (task_id, timestamp, source_agent, action, emotion, intensity, payload)
        VALUES (?, ?, ?, ?, ?, ?, ?)
      `).run(taskId, now, eventData.sourceAgent, eventData.action, eventData.emotion, eventData.intensity, eventData.payloadStr);

      // 2. update state
      if (stateChange) {
        db.prepare('UPDATE tasks SET bill_state = ?, updated_at = ? WHERE task_id = ?')
          .run(stateChange, now, taskId);
      }

      // 3. store act
      if (actJson) {
        db.prepare('INSERT OR REPLACE INTO acts (task_id, act_json, created_at) VALUES (?, ?, ?)')
          .run(taskId, actJson, now);
      }

      // 4. store verdict
      if (verdict) {
         db.prepare(`
          INSERT OR REPLACE INTO verdicts (task_id, constitutional, ruling, evidence, created_at)
          VALUES (?, ?, ?, ?, ?)
        `).run(taskId, verdict.constitutional ? 1 : 0, verdict.ruling, verdict.evidence, now);
      }
    });

    tx();
  }

  async getTaskEvents(
    taskId: string,
  ): Promise<Array<{ action: string; source_agent: string; payload: string }>> {
    const db = this.ensureDb();
    const rows = db.prepare(
      'SELECT action, source_agent, payload FROM events WHERE task_id = ? ORDER BY id ASC',
    ).all(taskId) as Array<{ action: string; source_agent: string; payload: string }>;
    return rows;
  }

  // ─── Acts ──────────────────────────────────────────────────────

  async storeAct(taskId: string, actJson: string): Promise<void> {
    const db = this.ensureDb();
    const now = new Date().toISOString();

    db.prepare(`
      INSERT OR REPLACE INTO acts (task_id, act_json, created_at) VALUES (?, ?, ?)
    `).run(taskId, actJson, now);
  }

  async getTaskAct(
    taskId: string,
  ): Promise<{ act_json: string; created_at: string } | null> {
    const db = this.ensureDb();
    const row = db.prepare('SELECT act_json, created_at FROM acts WHERE task_id = ?').get(taskId) as
      | { act_json: string; created_at: string }
      | undefined;
    return row ?? null;
  }

  // ─── Verdicts ──────────────────────────────────────────────────

  async storeVerdict(
    taskId: string,
    constitutional: boolean,
    ruling: string,
    evidence: string,
  ): Promise<void> {
    const db = this.ensureDb();
    const now = new Date().toISOString();

    // SQLite 没有 BOOLEAN 类型，存为 0/1 INTEGER
    db.prepare(`
      INSERT OR REPLACE INTO verdicts (task_id, constitutional, ruling, evidence, created_at)
      VALUES (?, ?, ?, ?, ?)
    `).run(taskId, constitutional ? 1 : 0, ruling, evidence, now);
  }

  async getTaskVerdict(
    taskId: string,
  ): Promise<{
    constitutional: number;
    ruling: string;
    evidence: string;
    created_at: string;
  } | null> {
    const db = this.ensureDb();
    const row = db.prepare(
      'SELECT constitutional, ruling, evidence, created_at FROM verdicts WHERE task_id = ?',
    ).get(taskId) as
      | { constitutional: number; ruling: string; evidence: string; created_at: string }
      | undefined;
    return row ?? null;
  }
}

File: openclaw_republic/server/task_store.py
"""SQLite 任务持久化存储。"""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Any

import aiosqlite
from pydantic import BaseModel

logger = logging.getLogger(__name__)


class TaskStatus(str, Enum):
    """任务状态。"""
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class TaskRecord(BaseModel):
    """任务记录。"""
    task_id: str
    petition: str
    status: TaskStatus
    result: str | None = None
    bill_state: str = "petition"
    created_at: datetime
    updated_at: datetime


class TaskStore:
    """SQLite 任务持久化存储。

    开发阶段使用 SQLite (aiosqlite)，后续可扩展 Redis。
    """

    def __init__(self, db_path: Path | str) -> None:
        """初始化 TaskStore。

        Args:
            db_path: SQLite 数据库文件路径。
        """
        self.db_path = Path(db_path)
        self._conn: aiosqlite.Connection | None = None

    async def initialize(self) -> None:
        """初始化数据库连接和表结构。"""
        # Ensure parent directory exists
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        
        self._conn = await aiosqlite.connect(self.db_path)
        # Enable foreign keys and set row factory
        await self._conn.execute("PRAGMA foreign_keys = ON")
        self._conn.row_factory = aiosqlite.Row

        # Create tables if not exist
        query_tasks = """
        CREATE TABLE IF NOT EXISTS tasks (
            task_id TEXT PRIMARY KEY,
            petition TEXT NOT NULL,
            status TEXT NOT NULL,
            result TEXT,
            bill_state TEXT NOT NULL DEFAULT 'petition',
            created_at TIMESTAMP NOT NULL,
            updated_at TIMESTAMP NOT NULL
        )
        """
        query_events = """
        CREATE TABLE IF NOT EXISTS events (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            task_id TEXT NOT NULL,
            timestamp TEXT NOT NULL,
            source_agent TEXT NOT NULL,
            action TEXT NOT NULL,
            emotion TEXT DEFAULT 'neutral',
            intensity REAL DEFAULT 0.5,
            payload TEXT DEFAULT '{}',
            FOREIGN KEY (task_id) REFERENCES tasks(task_id)
        )
        """
        query_acts = """
        CREATE TABLE IF NOT EXISTS acts (
            task_id TEXT PRIMARY KEY,
            act_json TEXT NOT NULL,
            created_at TEXT NOT NULL,
            FOREIGN KEY (task_id) REFERENCES tasks(task_id)
        )
        """
        query_verdicts = """
        CREATE TABLE IF NOT EXISTS verdicts (
            task_id TEXT PRIMARY KEY,
            constitutional BOOLEAN NOT NULL,
            ruling TEXT NOT NULL,
            evidence TEXT DEFAULT '[]',
            created_at TEXT NOT NULL,
            FOREIGN KEY (task_id) REFERENCES tasks(task_id)
        )
        """
        await self._conn.execute(query_tasks)
        await self._conn.execute(query_events)
        await self._conn.execute(query_acts)
        await self._conn.execute(query_verdicts)
        await self._conn.commit()
        logger.info("TaskStore initialized at %s", self.db_path)

    def _ensure_conn(self) -> aiosqlite.Connection:
        if not self._conn:
            raise RuntimeError("TaskStore not initialized. Call initialize() first.")
        return self._conn

    async def create_task(self, task_id: str, petition: str) -> TaskRecord:
        """创建新任务记录。

        Args:
            task_id: 任务 ID。
            petition: 请愿内容。

        Returns:
            创建的 TaskRecord。
        """
        conn = self._ensure_conn()
        now = datetime.now(timezone.utc)
        record = TaskRecord(
            task_id=task_id,
            petition=petition,
            status=TaskStatus.PENDING,
            bill_state="petition",
            created_at=now,
            updated_at=now,
        )

        query = """
        INSERT INTO tasks (task_id, petition, status, result, bill_state, created_at, updated_at)
        VALUES (?, ?, ?, ?, ?, ?, ?)
        """
        await conn.execute(
            query,
            (
                record.task_id,
                record.petition,
                record.status.value,
                record.result,
                record.bill_state,
                record.created_at.isoformat(),
                record.updated_at.isoformat(),
            ),
        )
        await conn.commit()
        return record

    async def get_task(self, task_id: str) -> TaskRecord | None:
        """查询单个任务。

        Args:
            task_id: 任务 ID。

        Returns:
            TaskRecord 对象，不存在则返回 None。
        """
        conn = self._ensure_conn()
        query = "SELECT * FROM tasks WHERE task_id = ?"
        async with conn.execute(query, (task_id,)) as cursor:
            row = await cursor.fetchone()

        if not row:
            return None

        return TaskRecord(
            task_id=row["task_id"],
            petition=row["petition"],
            status=TaskStatus(row["status"]),
            result=row["result"],
            bill_state=row["bill_state"],
            created_at=datetime.fromisoformat(row["created_at"]),
            updated_at=datetime.fromisoformat(row["updated_at"]),
        )

    async def update_task(self, task_id: str, **kwargs: Any) -> None:
        """更新任务状态/结果。

        Args:
            task_id: 任务 ID。
            **kwargs: 需要更新的字段（如 status, result, bill_state）。
        """
        if not kwargs:
            return

        conn = self._ensure_conn()
        now = datetime.now(timezone.utc).isoformat()
        
        updates = []
        values = []
        for key, val in kwargs.items():
            updates.append(f"{key} = ?")
            if isinstance(val, Enum):
                values.append(val.value)
            else:
                values.append(val)
                
        updates.append("updated_at = ?")
        values.append(now)
        values.append(task_id)

        set_clause = ", ".join(updates)
        query = f"UPDATE tasks SET {set_clause} WHERE task_id = ?"
        
        await conn.execute(query, values)
        await conn.commit()

    async def count_tasks(self) -> int:
        """获取任务总数。"""
        conn = self._ensure_conn()
        query = "SELECT COUNT(*) FROM tasks"
        async with conn.execute(query) as cursor:
            row = await cursor.fetchone()
            if row:
                return int(row[0])
            return 0

    async def list_tasks(self, offset: int = 0, limit: int = 20) -> list[TaskRecord]:
        """分页查询任务列表。

        Args:
            offset: 偏移量。
            limit: 最大返回数量。

        Returns:
            任务列表（按更新时间倒序）。
        """
        conn = self._ensure_conn()
        query = "SELECT * FROM tasks ORDER BY updated_at DESC LIMIT ? OFFSET ?"
        
        records = []
        async with conn.execute(query, (limit, offset)) as cursor:
            async for row in cursor:
                records.append(
                    TaskRecord(
                        task_id=row["task_id"],
                        petition=row["petition"],
                        status=TaskStatus(row["status"]),
                        result=row["result"],
                        bill_state=row["bill_state"],
                        created_at=datetime.fromisoformat(row["created_at"]),
                        updated_at=datetime.fromisoformat(row["updated_at"]),
                    )
                )
        return records

    async def store_event(self, task_id: str, source_agent: str, action: str, 
                          emotion: str = "neutral", intensity: float = 0.5, payload: str = "{}") -> None:
        """保存日志事件。"""
        conn = self._ensure_conn()
        now = datetime.now(timezone.utc).isoformat()
        query = """
        INSERT INTO events (task_id, timestamp, source_agent, action, emotion, intensity, payload)
        VALUES (?, ?, ?, ?, ?, ?, ?)
        """
        await conn.execute(query, (task_id, now, source_agent, action, emotion, intensity, payload))
        await conn.commit()

    async def get_task_events(self, task_id: str) -> list[aiosqlite.Row]:
        """获取任务的所有事件。"""
        conn = self._ensure_conn()
        query = "SELECT * FROM events WHERE task_id = ? ORDER BY id ASC"
        async with conn.execute(query, (task_id,)) as cursor:
            return list(await cursor.fetchall())

    async def store_act(self, task_id: str, act_json: str) -> None:
        """保存法案。"""
        conn = self._ensure_conn()
        now = datetime.now(timezone.utc).isoformat()
        query = """
        INSERT INTO acts (task_id, act_json, created_at)
        VALUES (?, ?, ?)
        """
        await conn.execute(query, (task_id, act_json, now))
        await conn.commit()

    async def get_task_act(self, task_id: str) -> aiosqlite.Row | None:
        """获取法案。"""
        conn = self._ensure_conn()
        query = "SELECT * FROM acts WHERE task_id = ?"
        async with conn.execute(query, (task_id,)) as cursor:
            return await cursor.fetchone()

    async def store_verdict(self, task_id: str, constitutional: bool, ruling: str, evidence: str = "[]") -> None:
        """保存判决。"""
        conn = self._ensure_conn()
        now = datetime.now(timezone.utc).isoformat()
        query = """
        INSERT INTO verdicts (task_id, constitutional, ruling, evidence, created_at)
        VALUES (?, ?, ?, ?, ?)
        """
        await conn.execute(query, (task_id, constitutional, ruling, evidence, now))
        await conn.commit()

    async def get_task_verdict(self, task_id: str) -> aiosqlite.Row | None:
        """获取判决。"""
        conn = self._ensure_conn()
        query = "SELECT * FROM verdicts WHERE task_id = ?"
        async with conn.execute(query, (task_id,)) as cursor:
            return await cursor.fetchone()

    async def close(self) -> None:
        """关闭数据库连接。"""
        if self._conn:
            await self._conn.close()
            self._conn = None
            logger.info("TaskStore connection closed.")
